File: Medium/Javascript/930. Binary Subarrays With Sum.js
/**Given a binary array nums and an integer goal, return the number of non-empty subarrays with a sum goal.
 * A subarray is a contiguous part of the array. */

/**
 * @param {number[]} nums
 * @param {number} goal
 * @return {number}
 */
var numSubarraysWithSum = function (nums, goal) {
  let map = {0 : 1};
  let ps = 0;
  let res = 0;

  for (let num of nums) {
    ps += num;
    let val = ps - goal;

    if (val in map) {
      res += map[val];
    }

    if (ps in map) {
      map[ps] += 1;
    } else {
      map[ps] = 1;
    }
  }

  return res;
};

console.log(numSubarraysWithSum([1,0,1,0,1],2));

File: Medium/python/930. Binary Subarrays With Sum.py
'''
Given a binary array nums and an integer goal, return the number of non-empty subarrays with a sum goal.

A subarray is a contiguous part of the array.
'''

class Solution:
    def numSubarraysWithSum(self, nums: list[int], goal: int) -> int:
        ps = 0
        mapping = { 0:1 }
        res = 0

        for num in nums:
            ps += num
            val = ps - goal
            if val in mapping:
                res += mapping[val]
            if ps in mapping:
                mapping[ps] += 1
            else:
                mapping[ps] = 1

        return res

print(Solution().numSubarraysWithSum([1, 1, 1, 1, 1],3))